.github/actions/publish-nuget/action.yml

```diff
@@ -0,0 +1,126 @@
+name: Publish to NuGet
+description: Builds all .NET solutions in the given directory with static code analysis.
+
+inputs:
+  source:
+    required: false
+    default: https://api.nuget.org/v3/index.json
+    description: The NuGet server URL used by the `dotnet nuget push` command's `--source` argument. 
+  verbosity:
+    required: false
+    default: minimal
+    description: The logging verbosity type used by the `dotnet` command.
+secrets:
+  API_KEY:
+    required: true
+
+runs:
+  using: "composite"
+  steps:
+    - name: Setup
+      id: setup
+      shell: pwsh
+      run: |
+        $source = @'
+        ${{ inputs.source }}
+        '@.Trim()
+
+        Write-Output "NuGet Source: $source"
+        "SOURCE_URL=$source" >> $Env:GITHUB_ENV
+        
+        $version = $Env:GITHUB_REF_NAME.Trim().TrimStart("v")
+        Write-Output "Target Version: $version"
+        "PUBLISH_VERSION=$version" >> $Env:GITHUB_ENV
+        Write-Output "::set-output name=publish-version::$version"
```

Comment:
Why do we need this both in an env var and output? It would be cleaner to just have the output. Same for `SOURCE_URL`.

Reply:
You need the output to be passed in the build step. I don't think replacing `$Env:PUBLISH_VERSION` with `'${{ steps.setup.outputs.publish-version }}'` is an improvement of clarity.

Reply:
The Build step already uses `${{ steps.setup.outputs.publish-version }}`. Other steps references `$Env:PUBLISH_VERSION` though. Why can't they also use the output instead?

For `SOURCE_URL`, you can just add an output parameter.

My issue with env vars here is that they're global variables while we have a mechanism to instead explicitly pass values between steps. While using them indeed requires that you locally change them, but for env vars you also have to muddle with your local environment or otherwise change how they're executed locally (and we're talking about one-liner scripts here that we rarely need to change or run locally, I'm not talking about the longer script files; those use parameters anyway).

Reply:
Ok, if you insist. I'm just very uncomfortable with this kind of jumping between different contexts and abstraction levels.
Also environment variables aren't actually global. They are action scoped.